protocol-specs/generators/dpos_random_seed_generation/index.js

```diff
@@ -0,0 +1,497 @@
+
+	return new Array(numberOfBlocks).fill(0).map((_v, index) => {
+		const height = startHeight + index;
+		const { publicKey } = delegateList[index % numberOfDelegates];
+		const seedReveal =
+			seedReveals[publicKey].seeds[seedReveals[publicKey].counter];
+
+		seedReveals[publicKey].counter += 1;
+
+		return {
+			generatorPublicKey: delegateList[index % numberOfDelegates].publicKey,
+			height,
+			seedReveal: seedReveal.toString('hex'),
+		};
+	});
+};
+
+const calcRound = (height, blocksPerRound) =>
+	Math.ceil(height / blocksPerRound);
+
+const roundInfo = (round, blocksPerRound) => {
+	const firstBlockHeight = round * blocksPerRound - blocksPerRound + 1;
+	const lastBlockHeight = round * blocksPerRound;
+	const middleBlockHeight = Math.floor(
+		(firstBlockHeight + lastBlockHeight) / 2,
+	);
+
+	return {
+		firstBlockHeight,
+		lastBlockHeight,
+		middleBlockHeight,
+	};
+};
+
+const findPreviousBlockOfDelegate = (block, searchTillHeight, blocksMap) => {
+	const { height, generatorPublicKey } = block;
+	for (let i = height - 1; i >= searchTillHeight; i -= 1) {
+		if (blocksMap[i].generatorPublicKey === generatorPublicKey) {
+			return blocksMap[i];
+		}
+	}
+
+	return null;
+};
+
+const isValidSeedReveal = (block, searchTillHeight, blocksMap) => {
+	const { height, seedReveal, generatorPublicKey } = block;
+
+	for (let i = height - 1; i >= searchTillHeight; i -= 1) {
+		if (blocksMap[i].generatorPublicKey === generatorPublicKey) {
+			return (
+				strippedHash(hexStrToBuffer(seedReveal)).toString('hex') !==
+				blocksMap[i].seedReveal
+			);
+		}
+	}
+
+	return false;
+};
+
+const generateRandomSeed = (blocks, blocksPerRound) => {
+	// Middle range of a round to validate
+	const middleThreshold = Math.floor(blocksPerRound / 2);
+	const lastBlockHeight = blocks[blocks.length - 1].height;
+	const currentRound = calcRound(lastBlockHeight, blocksPerRound);
+	const {
+		firstBlockHeight: firstBlockHeightOfCurrentRound,
+		middleBlockHeight: middleBlockHeightOfCurrentRound,
+	} = roundInfo(currentRound, blocksPerRound);
+	const { firstBlockHeight: firstBlockHeightOfLastRound } = roundInfo(
+		currentRound - 1,
+		blocksPerRound,
+	);
+	const { firstBlockHeight: firstBlockHeightOfSecondLastRound } = roundInfo(
+		currentRound - 2,
+		blocksPerRound,
+	);
+
+	/**
+	 * We need to build a map for current and last two rounds. To previously forged
+	 * blocks we will use only current and last round. To validate seed reveal of
+	 * any block from last round we have to load second last round as well.
+	 */
+	const blocksMap = blocks.reduce((acc, block) => {
+		if (block.height >= firstBlockHeightOfSecondLastRound) {
+			acc[block.height] = block;
+		}
+		return acc;
+	}, {});
+
+	if (lastBlockHeight < middleBlockHeightOfCurrentRound) {
+		throw new Error(
+			`Random seed can't be calculated earlier in a round. Wait till you pass middle of round. Current height ${lastBlockHeight}`,
+		);
+	}
+
+	if (currentRound === 1) {
```

Comment:
Any specification for first round in LIP, or this is something that came up during the discussion? @IkerAlus

Reply:
It came up after discussion, but it's exactly as per rules defined in LIP. If you run those rules for the first round, it will always have this particular output. So why I just excluded this condition for the first round.